File: monitor.js
const axios = require('axios');
const simpleGit = require('simple-git');
const fs = require('fs-extra');
const path = require('path');

// Configurações
const ACTUATOR_URL = 'http://localhost:8080/actuator/mappings'; // URL para consultar os endpoints
const GIT_REPO_URL = 'https://github.com/seu-usuario/repo-monitoramento.git'; // Repositório do GitHub
const REPO_DIR = path.join(process.cwd(), 'repo-monitoramento'); // Diretório do repositório clonado
const MONITOR_DIR = path.join(REPO_DIR, 'monitor'); // Pasta onde os arquivos serão salvos
const BRANCH_NAME = 'feature/create-alarms'; // Nome da nova branch

// Template de alerta
const ALERT_TEMPLATE = (endpoint, tags) => ({
  name: `Alert for ${endpoint}`,
  type: 'query alert',
  query: `avg(last_5m):avg:http.endpoint.latency{endpoint=${endpoint}} > 500`,
  message: `High latency detected for ${endpoint}`,
  tags: tags,
  priority: 'normal',
});

// Função para consultar os endpoints do Actuator
async function fetchEndpoints() {
  try {
    const response = await axios.get(ACTUATOR_URL);
    const endpoints = response.data.contexts.application.mappings.dispatcherServlets.dispatcherServlet;
    console.log(`Encontrados ${endpoints.length} endpoints.`);
    return endpoints;
  } catch (error) {
    console.error('Erro ao consultar os endpoints:', error.message);
    throw error;
  }
}

// Função para gerar arquivos JSON para cada endpoint
async function generateAlertFiles(endpoints) {
  if (!fs.existsSync(MONITOR_DIR)) {
    await fs.mkdirp(MONITOR_DIR);
  }

  for (const endpoint of endpoints) {
    const predicate = endpoint.predicate || '';
    const handler = endpoint.handler.split('#')[0].replace(/\./g, '_');
    const tags = [`endpoint:${predicate}`, 'service:my-service'];

    // Cria o JSON de alerta
    const alert = ALERT_TEMPLATE(predicate, tags);

    // Salva o arquivo na pasta monitor
    const filePath = path.join(MONITOR_DIR, `${handler}.json`);
    await fs.writeJson(filePath, alert, { spaces: 2 });
    console.log(`Arquivo criado: ${filePath}`);
  }
}

// Função para clonar o repositório
async function cloneRepository() {
  const git = simpleGit();

  if (fs.existsSync(REPO_DIR)) {
    console.log('Repositório já clonado. Usando o diretório existente.');
    return git.cwd(REPO_DIR);
  }

  console.log('Clonando o repositório...');
  await git.clone(GIT_REPO_URL, REPO_DIR);
  return git.cwd(REPO_DIR);
}

// Função para criar branch, commit e push
async function commitAndPushChanges(git) {
  console.log('Criando nova branch...');
  await git.checkoutLocalBranch(BRANCH_NAME);

  console.log('Adicionando arquivos...');
  await git.add('./*');

  console.log('Realizando commit...');
  await git.commit('feat: created alarms for all endpoints');

  console.log('Fazendo push...');
  await git.push('origin', BRANCH_NAME);
  console.log(`Alterações enviadas para a branch ${BRANCH_NAME}`);
}

// Função principal
async function main() {
  try {
    // 1. Consulta os endpoints
    const endpoints = await fetchEndpoints();

    // 2. Clona o repositório
    const git = await cloneRepository();

    // 3. Gera arquivos JSON
    await generateAlertFiles(endpoints);

    // 4. Commit e push das alterações
    await commitAndPushChanges(git);

    console.log('Processo concluído com sucesso!');
  } catch (error) {
    console.error('Erro durante o processo:', error.message);
  }
}

main();

File: monitor.py
import os
import json
import requests
from git import Repo

# Configurações
ACTUATOR_URL = "http://target-application:8080/actuator/mappings"
OUTPUT_DIR = "./alerts/"
GIT_REPO_URL = "https://github.com/seu-usuario/repo-alertas.git"
BRANCH_NAME = "auto-alerts-branch"

# Função para consultar os endpoints
def fetch_endpoints(url):
    response = requests.get(url)
    if response.status_code == 200:
        print("Endpoints obtidos com sucesso!")
        data = response.json()
        return data["contexts"]["application"]["mappings"]["dispatcherServlets"]["dispatcherServlet"]
    else:
        print(f"Erro ao consultar os endpoints: {response.status_code}")
        return []

# Função para gerar arquivos JSON
def generate_alert_files(endpoints, output_dir):
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    for endpoint in endpoints:
        predicate = endpoint.get("predicate", "")
        handler = endpoint.get("handler", "").split("#")[0].replace(".", "_")

        # Estrutura do alerta
        alert_data = {
            "name": f"Alert for {predicate}",
            "type": "query alert",
            "query": f"avg(last_5m):avg:http.endpoint.latency{{endpoint={predicate}}} > 500",
            "message": f"High latency detected for {predicate}",
            "priority": "normal",
        }

        # Salva o arquivo JSON
        file_path = os.path.join(output_dir, f"{handler}.json")
        with open(file_path, "w") as file:
            json.dump(alert_data, file, indent=4)

    print(f"Arquivos JSON gerados na pasta: {output_dir}")

# Função para clonar o repositório
def clone_repository(repo_url, local_path):
    if os.path.exists(local_path):
        print("Repositório já clonado. Usando a cópia local.")
        repo = Repo(local_path)
    else:
        print("Clonando o repositório...")
        repo = Repo.clone_from(repo_url, local_path)
    return repo

# Função para criar uma nova branch e fazer commit
def create_branch_and_commit(repo, branch_name, commit_message):
    # Cria uma nova branch
    if branch_name in repo.heads:
        repo.head.reference = repo.heads[branch_name]
    else:
        new_branch = repo.create_head(branch_name)
        repo.head.reference = new_branch

    repo.head.reset(index=True, working_tree=True)

    # Adiciona os arquivos e faz o commit
    repo.git.add(A=True)
    repo.index.commit(commit_message)
    print(f"Commit realizado na branch '{branch_name}'.")

# Função para push das alterações
def push_changes(repo, branch_name):
    try:
        origin = repo.remote(name="origin")
        origin.push(branch_name)
        print(f"Alterações enviadas para a branch '{branch_name}'.")
    except Exception as e:
        print(f"Erro ao enviar alterações: {e}")

# Função principal
def main():
    # Passo 1: Consultar os endpoints
    endpoints = fetch_endpoints(ACTUATOR_URL)

    # Passo 2: Gerar arquivos JSON
    generate_alert_files(endpoints, OUTPUT_DIR)

    # Passo 3: Clonar o repositório
    repo = clone_repository(GIT_REPO_URL, "./repo-alertas")

    # Passo 4: Criar nova branch e fazer commit
    create_branch_and_commit(repo, BRANCH_NAME, "Automated alerts for new endpoints")

    # Passo 5: Enviar alterações
    push_changes(repo, BRANCH_NAME)

if __name__ == "__main__":
    main()
